File: leetcode/reference/rotate_list.py
#!/usr/bin/env python3

class ListNode:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __repr__(self):
        return self.data

    def chain_string(self):
        node = self
        nodes = []
        while node is not None:
            nodes.append(str(node.data))
            node = node.next
        nodes.append("None")
        return " -> ".join(nodes)

class LinkedList:
    def __init__(self):
        self.head = None

    def __repr__(self):
        return self.head.chain_string()

def make_list(arr):
    llist = LinkedList()
    curr = None
    for item in arr:
        node = ListNode(item)
        if llist.head == None:
            llist.head = node
        else:
            curr.next = node
        curr = node
    return llist

class Solution:
    def rotateRight(self, head: 'ListNode', k: 'int') -> 'ListNode':
        # base cases
        if not head:
            return None
        if not head.next:
            return head

        # close the linked list into the ring
        old_tail = head
        n = 1
        while old_tail.next:
            old_tail = old_tail.next
            n += 1
        old_tail.next = head

        # find new tail : (n - k % n - 1)th node
        # and new head : (n - k % n)th node
        new_tail = head
        for i in range(n - k % n - 1):
            new_tail = new_tail.next
        new_head = new_tail.next

        # break the ring
        new_tail.next = None

        return new_head

arr = [1, 2, 3, 4, 5]
llist = make_list(arr)
print("llist = ", llist)
head = llist.head
k = 2
result = Solution().rotateRight(head, k)
print("result = ", result.chain_string())

File: leetcode/rotate_list.swift
#!/usr/bin/swift

class ListNode<T> {
    var value: T
    var next: ListNode?

    init(value: T) {
        self.value = value
    }
}

extension ListNode: CustomStringConvertible {
    var description: String {
        return "\(value)"
    }
}

extension ListNode {
    func chainString() -> String {
        var values = [String]()
        var cur: ListNode? = self
        while cur != nil {
            if let cur = cur {
                values.append("\(cur)")
            }
            cur = cur?.next
        }
        return values.joined(separator: " -> ")
    }
}

class LinkedList<T> {
    var head: ListNode<T>?
    init(head: ListNode<T>? = nil) {
        self.head = head
    }
}

extension LinkedList: CustomStringConvertible {
    var description: String {
        guard let head = head else {
            return "nil"
        }
        return head.chainString()
    }
}

func makeList<T>(from array: [T]) -> LinkedList<T> {
    let llist = LinkedList<T>()
    var current: ListNode<T>?
    for item in array {
        let node = ListNode<T>(value: item)
        if let current = current {
            current.next = node
        } else {
            llist.head = node
        }
        current = node
    }
    return llist
}

class Solution {
    func rotateRight(_ head: ListNode<Int>?, _ k: Int) -> ListNode<Int>? {
        guard let head = head else {
            return nil
        }
        if head.next == nil {
            return head
        }
        var oldTail = head
        var n = 1
        while oldTail.next != nil {
            guard let next = oldTail.next else {
                assertionFailure()
                break
            }
            oldTail = next
            n += 1
        }
        oldTail.next = head

        var newTail = head

        for _ in 0 ..< n - k % n - 1 {
            guard let next = newTail.next else {
                assertionFailure()
                break
            }
            newTail = next
        }
        guard let next = newTail.next else {
            assertionFailure()
            return nil
        }
        let newHead = next
        newTail.next = nil
        return newHead
    }
}

let arr = [1, 2, 3, 4, 5]
let llist = makeList(from: arr)
print("llist = \(llist)")
let head = llist.head
let k = 2
let result = Solution().rotateRight(head, k)
print("result = \(result?.chainString() ?? "nil")")
